File: core/x/javaapi/codegen/src/test/java/viaduct/x/javaapi/codegen/AccessorNameCollisionTest.java
package viaduct.x.javaapi.codegen;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static viaduct.x.javaapi.codegen.TestStrings.countOccurrences;

import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import org.junit.jupiter.api.Test;

/**
 * Regression tests for the accessor-name collision check on the Java GRT generator.
 *
 * <p>Without the check the generator emits both accessors and javac rejects the file for a
 * duplicate method the schema author never wrote.
 */
class AccessorNameCollisionTest {

  private static ObjectModel objectWith(List<FieldModel> fields) {
    return new ObjectModel("com.example.types", "Collision", List.of(), fields, null, false, false);
  }

  private static InterfaceModel interfaceWith(List<FieldModel> fields) {
    return new InterfaceModel("com.example.types", "Collision", List.of(), fields, null, false);
  }

  @Test
  void objectFieldCollidingWithStrictAccessorIsRejected() {
    ObjectModel model =
        objectWith(
            List.of(
                FieldModel.simple("foo", "String", true),
                FieldModel.simple("fooOrThrow", "String", true)));

    IllegalArgumentException error =
        assertThrows(
            IllegalArgumentException.class, () -> JavaGRTGenerator.ObjectGenerator.generate(model));

    assertTrue(error.getMessage().contains("type `Collision`"), error.getMessage());
    assertTrue(
        error.getMessage().contains("fields `foo` and `fooOrThrow` both generate `getFooOrThrow`"),
        error.getMessage());
  }

  @Test
  void interfaceFieldCollidingWithStrictAccessorIsRejected() {
    InterfaceModel model =
        interfaceWith(
            List.of(
                FieldModel.simple("foo", "String", true),
                FieldModel.simple("fooOrThrow", "String", true)));

    IllegalArgumentException error =
        assertThrows(
            IllegalArgumentException.class,
            () -> JavaGRTGenerator.InterfaceGenerator.generate(model));

    assertTrue(
        error.getMessage().contains("fields `foo` and `fooOrThrow` both generate `getFooOrThrow`"),
        error.getMessage());
  }

  /**
   * Java always prefixes {@code get}, so an {@code is} field collides under {@code
   * getIsReadyOrThrow}.
   */
  @Test
  void isPrefixedFieldCollidingWithStrictAccessorIsRejected() {
    ObjectModel model =
        objectWith(
            List.of(
                FieldModel.simple("isReady", "Boolean", true),
                FieldModel.simple("isReadyOrThrow", "Boolean", true)));

    IllegalArgumentException error =
        assertThrows(
            IllegalArgumentException.class, () -> JavaGRTGenerator.ObjectGenerator.generate(model));

    assertTrue(
        error
            .getMessage()
            .contains("fields `isReady` and `isReadyOrThrow` both generate `getIsReadyOrThrow`"),
        error.getMessage());
  }

  /** Two field names differing only in leading case map to one accessor. */
  @Test
  void caseOnlyCollisionIsRejected() {
    ObjectModel model =
        objectWith(
            List.of(
                FieldModel.simple("foo", "String", true),
                FieldModel.simple("Foo", "String", true)));

    IllegalArgumentException error =
        assertThrows(
            IllegalArgumentException.class, () -> JavaGRTGenerator.ObjectGenerator.generate(model));

    // The pair collides under every suffix, so the message names it once.
    assertTrue(
        error.getMessage().contains("fields `foo` and `Foo` both generate"), error.getMessage());
    assertEquals(1, countOccurrences(error.getMessage(), "both generate"), error.getMessage());
  }

  @Test
  void suffixNamedFieldWithoutSiblingIsGenerated() {
    ObjectModel model = objectWith(List.of(FieldModel.simple("fooOrThrow", "String", true)));

    String generated = assertDoesNotThrow(() -> JavaGRTGenerator.ObjectGenerator.generate(model));

    assertTrue(generated.contains("public String getFooOrThrowOrThrow()"), generated);
    assertTrue(generated.contains("public String getFooOrThrow()"), generated);
  }

  /** The suffix list comes from {@code AccessorForm}; this pins it to what the templates emit. */
  @Test
  void emittedAccessorNamesMatchTheSuffixListTheCheckUses() {
    ObjectModel model = objectWith(List.of(FieldModel.simple("foo", "String", true)));

    String generated = JavaGRTGenerator.ObjectGenerator.generate(model);

    Set<String> emitted =
        Pattern.compile("public String (getFoo\\w*)\\(")
            .matcher(generated)
            .results()
            .map(m -> m.group(1))
            .collect(Collectors.toSet());
    assertEquals(
        JavaGRTGenerator.FIELD_ACCESSOR_SUFFIXES.stream()
            .map(s -> "getFoo" + s)
            .collect(Collectors.toSet()),
        emitted);
  }

  @Test
  void fieldsDifferingOnlyByOrNullSuffixAreGenerated() {
    ObjectModel model =
        objectWith(
            List.of(
                FieldModel.simple("bar", "String", true),
                FieldModel.simple("barOrNull", "String", true)));

    String generated = assertDoesNotThrow(() -> JavaGRTGenerator.ObjectGenerator.generate(model));

    assertTrue(generated.contains("public String getBar()"), generated);
    assertTrue(generated.contains("public String getBarOrNull()"), generated);
  }
}

File: core/tenant/codegen/src/test/kotlin/viaduct/tenant/codegen/bytecode/testschema/AccessorNameCollisionTest.kt
package viaduct.tenant.codegen.bytecode.testschema

import graphql.schema.idl.SchemaParser
import java.io.File
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import viaduct.graphql.schema.ViaductSchema
import viaduct.graphql.schema.graphqljava.extensions.fromTypeDefinitionRegistry
import viaduct.graphql.utils.DefaultSchemaFactory
import viaduct.tenant.codegen.bytecode.CodeGenArgs
import viaduct.tenant.codegen.bytecode.GRTClassFilesBuilder
import viaduct.tenant.codegen.bytecode.config.ViaductBaseTypeMapper
import viaduct.tenant.codegen.bytecode.config.cfg
import viaduct.utils.timer.Timer

/**
 * Regression tests for the accessor-name collision check on the bytecode GRT generator.
 *
 * Undetected, these schemas reach javassist and fail there with a `DuplicateMemberException` naming a
 * method the schema author never wrote. Generation runs through
 * [GRTClassFilesBuilder.buildClassLoader], so schemas expected to generate are checked against
 * javassist as well as against this check.
 */
class AccessorNameCollisionTest {
    private fun generate(sdl: String): ClassLoader {
        val sdlFile = File.createTempFile("collision-", ".graphqls").also {
            it.deleteOnExit()
            it.writeText(sdl)
        }
        val defaultSdl = DefaultSchemaFactory.getDefaultSDL(
            existingSDLFiles = listOf(sdlFile),
            includeNodeDefinition = DefaultSchemaFactory.IncludeNodeSchema.Always,
            includeNodeQueries = DefaultSchemaFactory.IncludeNodeSchema.Never
        )
        val schema = ViaductSchema.fromTypeDefinitionRegistry(SchemaParser().parse(sdl + "\n" + defaultSdl))

        val args = CodeGenArgs(
            moduleName = null,
            pkgForGeneratedClasses = "viaduct.api.grts",
            includeIneligibleTypesForTestingOnly = false,
            excludeCrossModuleFields = false,
            javaTargetVersion = null,
            workerNumber = 0,
            workerCount = 1,
            timer = Timer(),
            baseTypeMapper = ViaductBaseTypeMapper(schema),
        )
        return GRTClassFilesBuilder(args).addAll(schema).buildClassLoader()
    }

    private fun assertRejects(
        sdl: String,
        expectedInMessage: String
    ) {
        val error = assertThrows<IllegalArgumentException> { generate(sdl) }
        assertTrue(
            error.message!!.contains(expectedInMessage),
            "Expected the collision to be reported as `$expectedInMessage`, got: ${error.message}"
        )
    }

    @Test
    fun `object field colliding with a strict accessor is rejected`() {
        assertRejects(
            """
            type Query { collision: Collision }
            type Collision { foo: String fooOrThrow: String }
            """.trimIndent(),
            "fields `foo` and `fooOrThrow` both generate `getFooOrThrow`"
        )
    }

    @Test
    fun `interface field colliding with a strict accessor is rejected`() {
        assertRejects(
            """
            type Query { collision: Collision }
            interface Collision { foo: String fooOrThrow: String }
            """.trimIndent(),
            "fields `foo` and `fooOrThrow` both generate `getFooOrThrow`"
        )
    }

    /** `Sub` would declare `getFooOrThrow()` on top of the one it inherits from `Base` for `foo`. */
    @Test
    fun `interface field colliding with an inherited accessor is rejected`() {
        assertRejects(
            """
            type Query { collision: Sub }
            interface Base { foo: String }
            interface Sub implements Base { foo: String fooOrThrow: String }
            """.trimIndent(),
            "fields `foo` and `fooOrThrow` both generate `getFooOrThrow`"
        )
    }

    /** An `is`-prefixed field keeps its name as its accessor, so neither side gets a `get` prefix. */
    @Test
    fun `is-prefixed field colliding with a strict accessor is rejected`() {
        assertRejects(
            """
            type Query { collision: Collision }
            type Collision { isReady: Boolean isReadyOrThrow: Boolean }
            """.trimIndent(),
            "fields `isReady` and `isReadyOrThrow` both generate `isReadyOrThrow`"
        )
    }

    @Test
    fun `object fields differing only by an OrNull suffix are generated`() {
        assertDoesNotThrow {
            generate(
                """
                type Query { collision: Collision }
                type Collision { bar: String barOrNull: String }
                """.trimIndent()
            )
        }
    }

    @Test
    fun `suffix-named fields without siblings are generated`() {
        assertDoesNotThrow {
            generate(
                """
                type Query { fine: Fine }
                type Fine { fooOrThrow: String barOrNull: String }
                """.trimIndent()
            )
        }
    }

    @Test
    fun `ordinary fields are generated`() {
        assertDoesNotThrow {
            generate(
                """
                type Query { fine: Fine }
                type Fine { foo: String bar: String isReady: Boolean }
                """.trimIndent()
            )
        }
    }

    /** Guards that a new `AccessorForm` cannot reach the emit sites without the check seeing it. */
    @Test
    fun `emitted accessor names match the suffix list the check uses`() {
        val classLoader = generate(
            """
            type Query { fine: Fine }
            type Fine { foo: String }
            """.trimIndent()
        )

        val emitted = classLoader.loadClass("viaduct.api.grts.Fine").declaredMethods
            .map { it.name }
            .filter { it.startsWith("getFoo") }
            .toSet()
        assertEquals(cfg.FIELD_ACCESSOR_SUFFIXES.map { "getFoo$it" }.toSet(), emitted)
    }
}
